### PR TITLE
Preserve metadata when reuploading files

### PR DESCRIPTION
**Pull request by gapra-msft**

### CODE REVIEW

File: component/azstorage/azstorage.go [hunk 1]
```diff
@@ -352,7 +352,7 @@ func (az *AzStorage) ReadInBuffer(options internal.ReadInBufferOptions) (length
 }
 
 func (az *AzStorage) WriteFile(options internal.WriteFileOptions) (int, error) {
-	err := az.storage.Write(options.Handle.Path, options.Offset, int64(len(options.Data)), options.Data, options.FileOffsets, options.ModBlockList)
+	err := az.storage.Write(options.Handle.Path, options.Offset, int64(len(options.Data)), options.Data, options.FileOffsets, options.ModBlockList, options.Metadata)
```
Comment (vibhansa-msft) [inline component/azstorage/azstorage.go:355]: From options we are passing down everything then why not pass the option structure itself ?

File: component/azstorage/azstorage.go [hunk 2]
```diff
@@ -387,7 +387,7 @@ func (az *AzStorage) TruncateFile(options internal.TruncateFileOptions) error {
 		}
 	}
 
-	err = az.storage.WriteFromBuffer(options.Name, nil, data)
+	err = az.storage.WriteFromBuffer(options.Name, attr.Metadata, data)
```
Comment (vibhansa-msft) [inline component/azstorage/azstorage.go:390]: If attr-cache is not in the pipeline, it shall come down as nil. We may need to make changes in libfuse layer for this.

**Review by vibhansa-msft — APPROVED** [review]

File: component/azstorage/azstorage.go (hunk 1 above)
Comment (gapra-msft) [reply component/azstorage/azstorage.go:355]: Yeah that is true. This is just how it was before. I can change it

File: component/azstorage/azstorage.go (hunk 2 above)
Comment (gapra-msft) [reply component/azstorage/azstorage.go:390]: Is there a problem with it being nil? We were passing nil before. The GO SDK can handle it

Comment (vibhansa-msft) [reply component/azstorage/azstorage.go:390]: No what I mean here is shall we explicitly pass nil where this call originates like libfuse or in file-cache or we want to rely on the default value being nil

File: component/azstorage/block_blob.go [hunk 3]
```diff
@@ -753,7 +753,7 @@ func (bb *BlockBlob) Write(name string, offset, length int64, data []byte, fileO
 			}
 		}
 		// WriteFromBuffer should be able to handle the case where now the block is too big and gets split into multiple blocks
-		err := bb.WriteFromBuffer(name, nil, *dataBuffer)
+		err := bb.WriteFromBuffer(name, metadata, *dataBuffer)
```
Comment (vibhansa-msft) [inline component/azstorage/block_blob.go:756]: In stage block and put block list also we need to take care of this metadata update?

**Review by tasherif-msft — APPROVED** [review]

**Comment by tasherif-msft** [comment]
how does the `noSyminks` option come into play with these changes?

File: component/azstorage/azstorage.go (hunk 2 above)
Comment (gapra-msft) [reply component/azstorage/azstorage.go:390]: Oh I see, we can pass nil thats fine

File: component/attr_cache/attr_cache.go [hunk 4]
```diff
@@ -349,6 +349,16 @@ func (ac *AttrCache) RenameFile(options internal.RenameFileOptions) error {
 
 // WriteFile : Mark the file invalid
 func (ac *AttrCache) WriteFile(options internal.WriteFileOptions) (int, error) {
+
+	// GetAttr on cache hit will serve from cache, on cache miss will serve from next component.
+	attr, err := ac.GetAttr(internal.GetAttrOptions{Name: options.Handle.Path, RetrieveMetadata: true})
+	if err != nil {
```
Comment (vibhansa-msft) [inline component/attr_cache/attr_cache.go:355]: If file does not exists in container then ignore the error.

Comment (gapra-msft) [reply component/attr_cache/attr_cache.go:356]: Can you review this once before I merge?

File: component/azstorage/azstorage.go [hunk 5]
```diff
@@ -352,7 +352,7 @@ func (az *AzStorage) ReadInBuffer(options internal.ReadInBufferOptions) (length
 }
 
 func (az *AzStorage) WriteFile(options internal.WriteFileOptions) (int, error) {
-	err := az.storage.Write(options.Handle.Path, options.Offset, int64(len(options.Data)), options.Data, options.FileOffsets, options.ModBlockList)
+	err := az.storage.Write(options)
```
Comment (gapra-msft) [inline component/azstorage/azstorage.go:355]: Can you review this too? @vibhansa-msft

File: component/azstorage/block_blob.go [hunk 6]
```diff
@@ -709,12 +709,15 @@ func (bb *BlockBlob) createNewBlocks(blockList *common.BlockOffsetList, offset,
 }
 
 // Write : write data at given offset to a blob
-func (bb *BlockBlob) Write(name string, offset, length int64, data []byte, fileOffsets, modFileOffsets *common.BlockOffsetList) error {
-	defer log.TimeTrack(time.Now(), "BlockBlob::Write", name)
+func (bb *BlockBlob) Write(options internal.WriteFileOptions) error {
+	name := options.Handle.Path
```
Comment (gapra-msft) [inline component/azstorage/block_blob.go:713]: modFileOffsets isnt used in this method, not sure why it was passed in in the first place

File: component/azstorage/azstorage.go (hunk 5 above)
Comment (vibhansa-msft) [reply component/azstorage/azstorage.go:355]: looks good (both the changes getattr and passing params)

**Review by vibhansa-msft — APPROVED** [review]

**Pull request merged**